modules/social_features/social_core/src/Controller/SecretFileController.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\social_core\Controller;
+
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Component\Utility\Crypt;
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Site\Settings;
+use Drupal\Core\StreamWrapper\StreamWrapperManagerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * Handles serving of requests using the secret file system.
+ */
+class SecretFileController extends ControllerBase {
+
+  /**
+   * The stream wrapper manager.
+   */
+  protected StreamWrapperManagerInterface $streamWrapperManager;
+
+   /**
+    * The Drupal time service.
+    */
+  protected TimeInterface $time;
+
+  /**
+   * FileDownloadController constructor.
+   *
+   * @param \Drupal\Core\StreamWrapper\StreamWrapperManagerInterface $stream_wrapper_manager
+   *   The stream wrapper manager.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The Drupal time service.
+   */
+  public function __construct(
+    StreamWrapperManagerInterface $stream_wrapper_manager,
+    TimeInterface $time
+  ) {
+    $this->streamWrapperManager = $stream_wrapper_manager;
+    $this->time = $time;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) : self {
+    return new static(
+      $container->get('stream_wrapper_manager'),
+      $container->get("datetime.time"),
+    );
+  }
+
+  /**
+   * Handles secret file transfers.
+   *
+   * We assume that if someone has a valid link then they can access the file
+   * that's being served by that link. The generators of the link should provide
+   * sufficient access checking.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request object.
+   *
+   * @return \Symfony\Component\HttpFoundation\BinaryFileResponse
+   *   The transferred file as response.
+   *
+   * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException
+   *   Thrown when the requested file does not exist or the link is invalid.
+   */
+  public function download(Request $request) : BinaryFileResponse {
+    $hash = $request->attributes->get('hash');
+    $expires_at = $request->attributes->get('expires_at');
+    $target = $request->attributes->get('filepath');
+    // Merge remaining path arguments into relative file path.
+    $uri = 'secret://' . $target;
+
+    // The secure hash may not be tampered with to prove the other variables in
+    // the URL are valid.
+    if (!self::confirmHash($expires_at, $target, $hash)) {
+      throw new NotFoundHttpException();
+    }
+
+    // The link must not have expired.
+    $time_left = $expires_at - $this->time->getRequestTime();
```

Comment:
```suggestion
    // @todo Since Drupal does not support adding cache metadata from
    // StreamWrappers it can currently occur that we cache a link in a render
    // cache longer than the link is valid which would cause viewers to be
    // provided with broken links. To avoid this we currently just assume that
    // a link with a valid hash is secret enough and thus always valid for some
    // time in the future. There is no current path to adding cache context in
    // StreamWrappers so there's no single issue to point to, but this is a
    // combination of various cache API and file system improvements that 
    // people are discussing. Once fixed, uncomment the next line and remove
    // the one after.
    // $time_left = $expires_at - $this->time->getRequestTime();
    $time_left = 3600 /* 1 hour */;
```

Reply:
As I understand we keep this at **1 hour** for the foreseeable future? What will be the impact change where typically static files like images are available for 6 months or longer?

Reply:
> I believe this is only the case for files that are directly served from the public filesystem by NGINX, which don't hit Drupal at all and are thus unaffected. The current private file system provides headers that prevent caching, so anything we move from private to secret will go from 0 -> 1 hour of caching.

At this point all files are, in our own case, part of the private file system. So nothing is currently cacheable.

I've been thinking about this a bit last night and I'm wondering if a fixed 1 hour will bring us the value we need. Or if we would need differentiation in for examples file types and visibility.

For example:
- 1 hour is OK for a high traffic site, because in 1 out of 100 requests, only a single one will be "uncached". Leading to cache hit ratio of 99%.
- 1 hour is OK, and needed, for files that are part of a closed group.
- 1 hour is also OK for any file that needs to be "downloaded", e.g. a PDF or XLS file. Something that doesn't render in the browser.

But on the other hand:
- Is 1 hour enough for a low traffic site, where the cache hit ratio is not 99%, but probably quite a bit less than 50% or in some cases 0%.
- Is the 1 hour not way too low for images that are served in public pages and even in pages that are part of the community that is open for anyone to sign up to? We could possibly increase it more than a month, ensuring us an average cache hit ratio again much closer to 100%.

And I raised it last night in the [comment below](https://github.com/goalgorilla/open_social/pull/3377#discussion_r1190334536) about the Drupal bootstrap, I do also think we need to do proper tests on how fast this system is when we're loading a page with 40 or more different images. What is the difference in speed in comparison to what we currently have?

I'm raising these points as I want to bring enough **value** considering we will be waiting, likely, for quite some time still until the core issue lands. And this is imho one of the easiest and probably the most impactful change we can still do on a short term.

Cc @ronaldtebrake, I would also like to have your input.

Reply:
> As I understand we keep this at 1 hour for the foreseeable future?

Correct. That's mostly to make sure that if we deploy the actual time-boxing in the future the old links won't sit in the CDN for a long time. The hour is chosen based on the idea that it sufficiently buckets requests together for busy sites while making links short-lived enough for security. If files aren't requested multiple times per hour then I assumed it's a low traffic site.

The hour was based on the default value in the StreamWrapper: https://github.com/goalgorilla/open_social/pull/3377/files/06f8467ce8230d0da228369ff9791ec185554a2c#diff-8cfbc562c225cd435052fe2f8ceb15bce6c742258f3bc461107c081d370d22e5R65

I've updated the suggestion to use the same setting instead which was already implemented there which allows tweaking on a per-site basis if we find we want a different caching/security trade-off.

> What will be the impact change where typically static files like images are available for 6 months or longer?
I believe this is only the case for files that are directly served from the public filesystem by NGINX, which don't hit Drupal at all and are thus unaffected. The current private file system provides headers that prevent caching, so anything we move from `private` to `secret` will go from 0 -> 1 hour of caching.

Reply:
@Kingdutch ping? :)

Reply:
We'll have to implement https://github.com/goalgorilla/cablecar/pull/367#issuecomment-1491680974
> In case the single Drupal bootstrap per time-interval per image is too much

Which takes care of this, I'll see if I can do that this week.